### PR TITLE
Adding Logs MetricFilter Dimensions and Unit Support

### PR DESCRIPTION
**Pull request by waifa**
*Issue #, if available:*
AWS::Logs::MetricFilter does not accept dimensions #64

*Description of changes:*
This PR is to implement a feature request for dimensions and unit support in Metric Filter via CFN. See https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-logs/issues/64 for more info.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-logs-metricfilter/aws-logs-metricfilter.json [hunk 1]
```diff
@@ -38,6 +61,22 @@
           "pattern": ".{1,100}",
           "minLength": 1,
           "maxLength": 100
+        },
+        "MetricUnit": {
```
Comment (evanuk) [inline aws-logs-metricfilter/aws-logs-metricfilter.json:65]: This should just be Unit, to match the api: https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutMetricFilter.html

File: aws-logs-metricfilter/aws-logs-metricfilter.json [hunk 2]
```diff
@@ -38,6 +61,22 @@
...
+          "description": "The unit to assign to the metric. If you omit this, the unit is set as None.",
+          "type": "string",
+          "enum":["Seconds", "Microseconds", "Milliseconds", "Bytes", "Kilobytes", "Megabytes", "Gigabytes", "Terabytes", "Bits", "Kilobits", "Megabits", "Gigabits", "Terabits", "Percent", "Count", "Bytes/Second", "Kilobytes/Second", "Megabytes/Second", "Gigabytes/Second", "Terabytes/Second", "Bits/Second", "Kilobits/Second", "Megabits/Second", "Gigabits/Second", "Terabits/Second", "Count/Second", "None"]
+        },
+        "MetricDimensions": {
```
Comment (evanuk) [inline aws-logs-metricfilter/aws-logs-metricfilter.json:70]: Same as above, just Dimensions

File: aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java [hunk 3]
```diff
@@ -33,29 +31,61 @@ else if (e instanceof ServiceUnavailableException) {
...
 
   static software.amazon.logs.metricfilter.MetricTransformation translateMetricTransformationFromSdk
-          (final software.amazon.awssdk.services.cloudwatchlogs.model.MetricTransformation metricTransformation) {
+  (final software.amazon.awssdk.services.cloudwatchlogs.model.MetricTransformation metricTransformation) {
     if (metricTransformation == null) {
       return null;
     }
-    return software.amazon.logs.metricfilter.MetricTransformation.builder()
+
+    software.amazon.logs.metricfilter.MetricTransformation returnObj = software.amazon.logs.metricfilter.MetricTransformation.builder()
```
Comment (wchunl) [inline aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java:67]: small np: we could probably use a better var name than `returnObj`

File: aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/AbstractTestBase.java [hunk 4]
```diff
@@ -1,8 +1,11 @@
 package software.amazon.logs.metricfilter;
 
-import java.util.Arrays;
+import java.util.*;
```
Comment (wchunl) [inline aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/AbstractTestBase.java:3]: np: we import `java.util.*` here then import `java.util.stream.Collectors`, may be better to just avoid using import *

File: aws-logs-metricfilter/aws-logs-metricfilter.json (hunk 1 above)
Comment (waifa) [reply aws-logs-metricfilter/aws-logs-metricfilter.json]: updated

File: aws-logs-metricfilter/aws-logs-metricfilter.json (hunk 2 above)
Comment (waifa) [reply aws-logs-metricfilter/aws-logs-metricfilter.json]: updated
good catch, thanks!

File: aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java [hunk 5]
```diff
@@ -33,29 +31,61 @@ else if (e instanceof ServiceUnavailableException) {
...
-    return software.amazon.awssdk.services.cloudwatchlogs.model.MetricTransformation.builder()
+
+    software.amazon.awssdk.services.cloudwatchlogs.model.MetricTransformation returnObj = software.amazon.awssdk.services.cloudwatchlogs.model.MetricTransformation.builder()
             .metricName(metricTransformation.getMetricName())
             .metricValue(metricTransformation.getMetricValue())
             .metricNamespace(metricTransformation.getMetricNamespace())
             .defaultValue(metricTransformation.getDefaultValue())
             .build();
+
+    if (metricTransformation.getMetricDimensions() != null && !metricTransformation.getMetricDimensions().isEmpty()){
```
[the remark this answers is not in the document]
Comment (waifa) [reply aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java:46]: The sdk model uses a hashmap and the logs model uses a set. My initial implementation did use the same structure as the other variables to create the dimension, but that doesn't work due to the type difference.

File: aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java [hunk 6]
```diff
@@ -33,29 +31,61 @@ else if (e instanceof ServiceUnavailableException) {
...
+      HashMap<String, String> dimensionsMap = new HashMap<String, String>();
+      for (Dimension entry: metricTransformation.getMetricDimensions()) {
+        dimensionsMap.put(entry.getKey(), entry.getValue());
+      }
+      returnObj = returnObj.toBuilder().dimensions(dimensionsMap).build();
```
[the remark this answers is not in the document]
Comment (waifa) [reply aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java:51]: so this follows the same reasoning as the comment above. The type returned by `getMetricDimensions()` logs model object is a set, so it needs to be converted to a hashmap.

File: aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java (hunk 3 above)
Comment (waifa) [reply aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java:67]: agreed, updating that

File: aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/AbstractTestBase.java (hunk 4 above)
Comment (waifa) [reply aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/AbstractTestBase.java:3]: agreed, updating imports

**Review by wchunl — APPROVED** [review]

File: aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java [hunk 7]
```diff
@@ -5,9 +5,7 @@
 import software.amazon.awssdk.services.cloudwatchlogs.model.ResourceNotFoundException;
 import software.amazon.cloudformation.exceptions.*;
 
-import java.util.Collection;
-import java.util.List;
-import java.util.Optional;
+import java.util.*;
```
Comment (evanuk) [inline aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java:8]: nit: can we avoid using wildcard imports

File: aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/TranslatorTest.java [hunk 8]
```diff
@@ -9,28 +9,37 @@
 import software.amazon.awssdk.services.cloudwatchlogs.model.MetricFilter;
 import software.amazon.awssdk.services.cloudwatchlogs.model.PutMetricFilterRequest;
 
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
```
Comment (evanuk) [inline aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/TranslatorTest.java:12]: Same as above, avoid using wildcard imports.

File: aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java (hunk 7 above)
Comment (waifa) [reply aws-logs-metricfilter/src/main/java/software/amazon/logs/metricfilter/Translator.java]: updated!

File: aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/TranslatorTest.java (hunk 8 above)
Comment (waifa) [reply aws-logs-metricfilter/src/test/java/software/amazon/logs/metricfilter/TranslatorTest.java]: updated!

**Review by evanuk — APPROVED** [review]

**Comment by waifa** [comment]
verifying that this PR successfully passes all unit tests and builds(via "mvn package")

**Comment by VRevol** [comment]
Does it mean it is from now on available to use in CFN ?
Documentation is still not referencing it ;)
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-logs-metricfilter-metrictransformation.html

I may give it a try ;)

File: aws-logs-metricfilter/aws-logs-metricfilter.json [hunk 9]
```diff
@@ -10,6 +10,29 @@
...
+    "Dimension": {
+      "description": "the key-value pairs that further define a metric.",
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "Key": {
+          "type": "string",
+          "description": "The key of the dimension. Maximum length of 255.",
+          "minLength": 1,
+          "maxLength": 255
```
Comment (markkuhn) [inline aws-logs-metricfilter/aws-logs-metricfilter.json:22]: Needs to be changed to 250 - see [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_structure_example)

**Pull request merged**